Feature/refactor react

dustinbyershax opened the pull request:
## Summary

Complete frontend modernization implementing React 19.2 + TypeScript with comprehensive Docker development environment and production deployment infrastructure.

Closes #13

## Frontend Implementation

### Core Architecture
- React 19.2 with TypeScript and strict type checking
- Vite build system with HMR for fast development
- React Router v7 for client-side routing
- TanStack Query v5 for server state management
- Tailwind CSS 4 with shadcn/ui component library

### Features & Components
- **Politician Search** - Real-time search with autocomplete, detailed profiles, voting history
- **Donor Search** - Search donors with contribution history and industry categorization
- **Vote Filtering** - Filter votes by bill type, subject, and date with multi-select UI
- **Data Visualization** - Chart.js integration for donation trends and contribution analysis
- **Politician Comparison** - Side-by-side comparison of voting records and funding sources
- **Command Palette** - Keyboard-driven navigation (Cmd+K) for quick access
- **Theme System** - Dark/light mode with persistent preferences
- **Error Boundaries** - Comprehensive error handling and user feedback

### Developer Experience
- ESLint 9 + TypeScript ESLint for code quality
- Prettier with Tailwind plugin for consistent formatting
- Vitest for unit testing with React Testing Library
- Knip for unused code detection
- 90+ npm packages organized in modular structure

## Backend Updates

### Flask Application
- Static file serving for React SPA with proper routing support
- CORS configuration with security-aware conditional enablement
- React Router integration - catch-all route serves index.html for client-side routing
- Boolean parsing fix for `FLASK_DEBUG` preventing CORS security misconfiguration

### Configuration Management
- Docker environment detection to use correct database host (`db` vs `localhost`)
- Sensible defaults for local development without `.env` file
- Removed dead code warning for database credentials (defaults intentional)
- Environment variable precedence: Docker Compose > `.env` > defaults

## Infrastructure

### Docker Compose Development
- **PostgreSQL 16** - Alpine-based with health checks and persistent volumes
- **Backend** - Python 3.13 with hot reload, mounted source directories
- **Frontend** - Node 24 with pnpm, Vite dev server on port 5173
- Automatic database schema initialization from `pg-dump.tar.bz2`
- Service dependencies and health checks for reliable startup

### Multi-Stage Production Dockerfile
- Stage 1: Node 24 Alpine builds optimized frontend bundle
- Stage 2: Python 3.13 Alpine serves backend + static frontend
- Built frontend copied to `app/static/` for single-container deployment
- Non-root user for security
- Gunicorn with 4 workers, 120s timeout

### Documentation
- Comprehensive README with Docker Compose quick start
- Database initialization and population script documentation
- Local development setup (with and without Docker)
- External data requirements for populate scripts
- SQL query examples and database access instructions

## Developer Tooling

### Scripts & Automation
- `bin/docker-entrypoint.sh` - Automatic schema restoration on first run
- `bin/start_local.sh` / `.bat` - Cross-platform local dev startup
- `bin/check_and_seed.py` - Database initialization verification
- `bin/init_db.py` - Schema restoration from backup
- Populate scripts for politicians, bills, donors, donations, votes, industries

### Configuration Files
- `.dockerignore` - Comprehensive exclusions for smaller images
- `.gitignore` - Frontend lib/ exception to include React source utilities
- `.prettierrc` / `.prettierignore` - Team code formatting standards
- `knip.config.ts` - Dead code detection configuration
- `eslint.config.js` - Modern flat config with React rules

## Bug Fixes & Security

- **Security**: Fixed `FLASK_DEBUG` boolean parsing - prevents CORS enabled in production
- **Code Quality**: Removed unused `API_BASE_URL` constant (empty string no-op)
- **Documentation**: Clarified database credential defaults are intentional
- **Tooling**: Tailwind data attribute fix script for component compatibility

## Testing

- Unit test framework configured (Vitest + React Testing Library)
- Sample test for `PoliticianCard` component with 191 lines of coverage
- Test scripts: `test`, `test:ui`, `test:run` for different workflows

---

## Known Issues (Created During Review)

Issues identified during code review that will be addressed in follow-up PRs:

### Production Blockers
- #30 - Add maximum query length validation to prevent DoS
- #31 - Fix error message leakage exposing internal details
- #32 - Fix Docker environment variable override logic
- #33 - Add frontend test coverage for critical user flows
- #35 - Add database connection pooling for better performance

### Improvements
- #34 - Replace print() statements with proper logging
- #36 - Remove console.log statements from production frontend code
- #37 - Add health check to production Docker container
- #38 - Document external data sources for database population

---

This PR transforms the application from inline HTML/CSS/JS to a modern, maintainable React application with professional development and deployment infrastructure.

dustinbyershax committed:
add-frontend

dustinbyershax committed:
mv and update README

dustinbyershax committed:
edit README. update main.py for react app support w/ react-router

dustinbyershax committed:
update local docker set and README

dustinbyershax committed:
update README

dustinbyershax committed:
Fix .gitignore to allow frontend/src/lib/ directory

- Add exception rule for frontend/src/lib/ to prevent it from being ignored
- The lib/ pattern in Python section was incorrectly ignoring frontend source code
- Add frontend/src/lib/ files to version control

eddyizm commented on app/main.py:
```diff
@@ -483,6 +470,17 @@ def get_all_bill_subjects():
             conn.close()
 
 
```
```suggestion
@app.route('/health', methods=['GET'])
def health_check():
    return jsonify({'status': 'healthy'}), 200


```
Since we don't have a full swagger ui for the endpoints, we can start here.

eddyizm commented on app/config.py:
```diff
@@ -2,19 +2,32 @@
+DB_NAME = os.getenv("DB_NAME") or "paper_trail_dev"
+DB_USER = os.getenv("DB_USER") or "paper_trail_user"
+DB_PASSWORD = os.getenv("DB_PASSWORD") or "dev_password_change_in_production"
 
```
```suggestion
DB_PORT = os.getenv("DB_PORT", "5432")
DB_NAME = os.getenv("DB_NAME", "paper_trail_dev")
DB_USER = os.getenv("DB_USER",  "paper_trail_user")
DB_PASSWORD = os.getenv("DB_PASSWORD", "dev_password_change_in_production")

FLASK_DEBUG = os.getenv("FLASK_DEBUG", None)
```
More conventional use of `getenv` along with the env var that tripped me up the most. This was already defined but there was a new one, `FLASK_ENV`, I propose sticking with this one as when this is rolled up to prod, i just don't include the debug key in the file so it never can go to debug.

eddyizm commented on app/config.py:
```diff
@@ -2,19 +2,32 @@
+    DB_HOST = "db"
+else:
+    # Use .env file value or default to localhost
+    DB_HOST = os.getenv("DB_HOST") or "localhost"
```
```suggestion
    DB_HOST = os.getenv("DB_HOST", "localhost") 
```

dustinbyershax committed:
Update Dockerfile

Co-authored-by: eddyizm <eddyizm@gmail.com>

dustinbyershax committed:
Update config.py

Co-authored-by: eddyizm <eddyizm@gmail.com>

dustinbyershax committed:
Update config.py

Co-authored-by: eddyizm <eddyizm@gmail.com>

dustinbyershax committed:
Update main.py

Co-authored-by: eddyizm <eddyizm@gmail.com>

dustinbyershax committed:
Update main.py

Co-authored-by: eddyizm <eddyizm@gmail.com>

dustinbyershax committed:
remove unneeded API_BASE_URL - URLs are already relative paths so empty string is a no-op effectively

dustinbyershax committed:
Remove dead code warning for database credentials

The sanity check warning for DB_NAME, DB_USER, and DB_PASSWORD was
never triggered because these variables have sensible defaults for
local development. Since the defaults are intentional, the warning
has been removed. The CONGRESS_GOV_API_KEY warning remains as it
has no default value.

dustinbyershax committed:
fix: Parse FLASK_DEBUG as boolean to prevent CORS security misconfiguration

FLASK_DEBUG was set to os.getenv() which returns a string. In Python, any
non-empty string is truthy, so FLASK_DEBUG="False" would enable CORS in
production (security vulnerability).

Changed to: FLASK_DEBUG = os.getenv("FLASK_DEBUG", "False").lower() == "true"

This properly parses the environment variable as a boolean:
- "true", "True", "TRUE" → True (enables CORS)
- "false", "False", "FALSE", unset, or any other value → False (disables CORS)

Prevents accidental CORS enablement in production environments.
